test/acceptance/js/FilestoreTests.js

```diff
@@ -320,14 +397,307 @@ describe('Filestore', function() {
+              const res = await rp.get(fileUrl)
+              expect(res.body).to.equal(constantFileContent)
+            })
+
+            describe('when copyOnMiss is disabled', function() {
+              beforeEach(function() {
+                Settings.filestore.fallback.copyOnMiss = false
+              })
+
+              it('should not copy the file to the primary', async function() {
+                await rp.get(fileUrl)
+
+                await expectPersistorNotToHaveFile(
+                  app.persistor.primaryPersistor,
+                  bucket,
+                  fileKey
+                )
+              })
+            })
+
+            describe('when copyOnMiss is enabled', function() {
+              beforeEach(function() {
+                Settings.filestore.fallback.copyOnMiss = true
+              })
+
+              it('copies the file to the primary', async function() {
+                await rp.get(fileUrl)
+                // wait for the file to copy in the background
+                await promisify(setTimeout)(1000)
+
+                await expectPersistorToHaveFile(
+                  app.persistor.primaryPersistor,
+                  bucket,
+                  fileKey,
+                  constantFileContent
+                )
+              })
+            })
+
+            describe('when copying a file', function() {
+              let newFileId, newFileUrl, newFileKey
+
+              beforeEach(async function() {
+                const newProjectID = `acceptance_tests_copied_project_${Math.random()}`
+                newFileId = Math.random()
+                newFileUrl = `${filestoreUrl}/project/${newProjectID}/file/${directoryName}%2F${newFileId}`
+                newFileKey = `${newProjectID}/${directoryName}/${newFileId}`
+
+                const opts = {
+                  method: 'put',
+                  uri: newFileUrl,
+                  json: {
+                    source: {
+                      project_id: projectId,
+                      file_id: `${directoryName}/${fileId}`
+                    }
+                  }
+                }
+
+                const response = await rp(opts)
+                expect(response.statusCode).to.equal(200)
+              })
+
+              it('should leave the old file in the old bucket', async function() {
+                await expectPersistorToHaveFile(
+                  app.persistor.fallbackPersistor,
+                  fallbackBucket,
+                  fileKey,
+                  constantFileContent
+                )
+              })
+
+              it('should not create a new file in the old bucket', async function() {
+                await expectPersistorNotToHaveFile(
+                  app.persistor.fallbackPersistor,
+                  fallbackBucket,
+                  newFileKey
+                )
+              })
+
+              it('should not copy the old file to the new bucket', async function() {
```

Comment:
What is the rational for that?

A followup copy call in the primary persistor would save us another download from the fallback + upload to the primary.

Reply:
Good question. If `copyOnMiss` is enabled it's probably a good idea to copy it to both locations. :+1: